### PR TITLE
Configurable builds

### PR DESCRIPTION
**Pull request by peterbe**
Fixes #966

Basically, it's now making it possible to specify specific files and folders when building. Let's include everything that https://github.com/Gregoor/yari/pull/4 brought:

* `flawLevels` - can now be included in `yarn build` and `yarn start` by env vars. 
* `files` - is now respected, but via an env var instead of a CLI flag. This makes the PR Builds CI work again (in theory)
* `foldersearch` - by far the least important feature but something I often find myself using when hacking on doing mass-builds but don't want to have to wait for all 10k docs
* `noProgressbar` - very useful in CI and since `CI=true` is a "standard" in CI environments, we should probably consider `process.env.BUILD_NO_PROGRESSBAR || process.env.CI`

### CODE REVIEW

File: build/build-options.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+const {
+  FLAW_LEVELS,
+  DEFAULT_FLAW_LEVELS,
+  VALID_FLAW_CHECKS,
+  FILES,
+  FOLDERSEARCH,
+  NO_PROGRESSBAR,
+} = require("./constants");
+
+module.exports = {
```
Comment (Gregoor) [inline build/build-options.js]: nit: I'd vote for moving it down to not rely on function hoisting

File: build/build-options.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  flawLevels: parseFlawLevels(DEFAULT_FLAW_LEVELS),
+  files: parseFiles(FILES),
+  foldersearch: parseFoldersearch(FOLDERSEARCH),
+  noProgressbar: NO_PROGRESSBAR,
```
Comment (Gregoor) [inline build/build-options.js:16]: Why the re-export?

File: content/document.js [hunk 3]
```diff
@@ -213,9 +213,29 @@ const findByURL = memoize((url, fields = null) => {
   return document ? { contentRoot: CONTENT_ROOT, folder, document } : null;
 });
 
-function findAll() {
+function findAll(
+  { files, foldersearch } = { files: new Set(), foldersearch: null }
+) {
+  if (!files instanceof Set) throw new Error("TypeError: 'file' not a Set");
+  if (foldersearch && typeof foldersearch !== "string")
+    throw new Error("TypeError: 'file' not a Set");
```
Comment (Gregoor) [inline content/document.js]: error message is off

File: content/document.js [hunk 4]
```diff
@@ -213,9 +213,29 @@ const findByURL = memoize((url, fields = null) => {
   return document ? { contentRoot: CONTENT_ROOT, folder, document } : null;
 });
 
-function findAll() {
+function findAll(
+  { files, foldersearch } = { files: new Set(), foldersearch: null }
```
Comment (Gregoor) [inline content/document.js]: nit: `foldersearch` casing is off. Maybe I'll also prefer something like `folderPattern` (though I guess that suggests that it already supports regex).

**Review by Gregoor — APPROVED** [review]
Thanks Peter! Only got some nits. Still not a fan of using env vars, but its my own fault for ripping CLI out of the build. Let's revisit in the future.

File: build/build-options.js (hunk 1 above)
Comment (peterbe) [reply build/build-options.js]: Got it.

File: build/build-options.js (hunk 2 above)
Comment (peterbe) [reply build/build-options.js:16]: I don't get it.

Comment (Gregoor) [reply build/build-options.js:16]: Yeah actually nevermind, I was wondering why `NO_PROGRESSBAR` isn't imported directly where it's needed. But then again, this current way makes it easier to replace with CLI options in the future, so I do like it more. Nevermind me 🙊

Comment (peterbe) [reply build/build-options.js:16]: I want whoever uses `require('./build-options')` to refer to it as `noProgressbar`.

File: content/document.js [hunk 5]
```diff
@@ -213,9 +213,29 @@ const findByURL = memoize((url, fields = null) => {
   return document ? { contentRoot: CONTENT_ROOT, folder, document } : null;
 });
 
-function findAll() {
+function findAll(
+  { files, folderSearch } = { files: new Set(), folderSearch: null }
+) {
+  if (!files instanceof Set) throw new Error("TypeError: 'files' not a Set");
```
Comment (Gregoor) [inline content/document.js]: oh another nit: why not throw a [real TypeError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError/TypeError)?

Comment (peterbe) [reply content/document.js]: done.

**Comment by peterbe** [comment]
> Still not a fan of using env vars, but its my own fault for ripping CLI out of the build. Let's revisit in the future.

Would you say that, if time wasn't an issue, that CLI would be better? (I think I know the answer and I don't think I disagree)

I think what I've come to accept is that it's annoying to have a bit of both. 

For running the dev server, we use `create-react-app` so you have to use env vars to shape the behaviors it in specific ways. 
But the mass-builds, they're really pretty rare in a sense since it's only ever used for the core dev team and for CI. So, with that in mind, I don't worry about the risk of 10s or 100s of contributors accidentally messing up typos in their `.env` file. 

After all, I think this is pretty good enough.

**Pull request merged**